### PR TITLE
Enable Tooltips for received one day view stacked bar graphs

### PR DESCRIPTION
**Pull request by DanielMwendwa**
Added tooltips for the received one day stacked bar graph. Look at this before I apply the tooltips to all the stacked bar graphs
Implementation: When one hovers on a bar graph I'll get the color of the bar and I'll use a function to look up the name of the operator that matches the color as defined in the variable `operator_identity`. I'll query the data for the given operator to output the number of messages on the tooltip.
tooltips

### CODE REVIEW

File: public/graph.js [hunk 1]
```diff
@@ -383,6 +383,50 @@ class GraphController {
...
+        function componentToHex(c) {
+            var hex = c.toString(16);
+                return hex.length == 1 ? "0" + hex : hex;
+            }
+
+        function rgbToHex(r, g, b) {
+            return "#" + componentToHex(r) + componentToHex(g) + componentToHex(b);
+        }
+
+        let operators_identity = {
```
Comment (IsaackMwenda) [inline public/graph.js]: let us use the mno_colors table available in avf-dashboards firebase.

**Review by IsaackMwenda — CHANGES_REQUESTED** [review]

File: public/graph.js [hunk 2]
```diff
@@ -499,6 +530,36 @@ class GraphController {
...
+                .selectAll("rect")
+                .on("mouseover", (d, i, n) => {
+                    // Get color of hovered rect
+                    let rgb = d3.select(n[i]).style("fill")
+                    // Get color components from an rgb string
+                    rgb = rgb.substring(4, rgb.length-1).replace(/ /g, '').split(',');
+                    // Cast strings in array to int
+                    rgb = rgb.map((x) =>parseInt(x));
+                    let hex = rgbToHex(...rgb)
+                    let op = legendColorToOperator(hex)
```
Comment (IsaackMwenda) [inline public/graph.js]: Change to a descriptive variable name, please.

File: public/styles.css [hunk 3]
```diff
@@ -197,3 +197,12 @@ table > tbody > tr > td {
 .alert-stale-info {
     background-color: #f08080;
 }
+.tooltip2 {
```
Comment (IsaackMwenda) [inline public/styles.css]:
```suggestion
.tooltip{	
```
?

File: public/graph.js [hunk 4]
```diff
@@ -499,6 +530,36 @@ class GraphController {
...
+                
+                    tip = d3.tip()
+                        .attr("class", "tooltip2")
+                        .attr("id", "tooltip2")
+                        .html(d => {
+                            let num = d.data[`${op}_received`]
+                            let content = `<div>${op} ${num}</div>` 
+                            return content;
+                    })
+
```
Comment (IsaackMwenda) [inline public/graph.js]:
```suggestion
```
remove unnecessary white spaces here and in the rest of the code

Comment (IsaackMwenda) [inline public/graph.js]:
```suggestion
```

File: public/graph.js [hunk 5]
```diff
@@ -499,6 +530,36 @@ class GraphController {
...
+                    // Cast strings in array to int
+                    rgb = rgb.map((x) =>parseInt(x));
+                    let hex = rgbToHex(...rgb)
+                    let op = legendColorToOperator(hex)
+                
+                    tip = d3.tip()
+                        .attr("class", "tooltip2")
+                        .attr("id", "tooltip2")
+                        .html(d => {
+                            let num = d.data[`${op}_received`]
```
Comment (IsaackMwenda) [inline public/graph.js]:
```suggestion
                            let total_receved_no = d.data[`${op}_received`]
```
? or better

File: public/graph.js [hunk 6]
```diff
@@ -499,6 +530,36 @@ class GraphController {
...
+                            let content = `<div>${op} ${num}</div>`
```
Comment (IsaackMwenda) [inline public/graph.js]:
```suggestion
                            let tooltip_content = `<div>${op} ${num}</div>`  
```
?

File: public/graph.js [hunk 7]
```diff
@@ -499,6 +530,36 @@ class GraphController {
...
+            receivedLayer
+                .selectAll("rect")
+                .on("mouseover", (d, i, n) => {
+                    // Get color of hovered rect
+                    let rgb = d3.select(n[i]).style("fill")
+                    // Get color components from an rgb string
+                    rgb = rgb.substring(4, rgb.length-1).replace(/ /g, '').split(',');
+                    // Cast strings in array to int
+                    rgb = rgb.map((x) =>parseInt(x));
+                    let hex = rgbToHex(...rgb)
```
Comment (IsaackMwenda) [inline public/graph.js]: If operator colors are already stored in hex format, we should probably pull this from the table instead?

**Review by IsaackMwenda — CHANGES_REQUESTED** [review]
Requesting changes because of a question re-color conversion

**Comment by DanielMwendwa** [comment]
Issue: having trouble accessing the relevant values of the initial dataset for a mousover tooltip

Reason: `d` is the Array[2] at this point with the values defining baseline / topline, `d.data` is the complete original data object but is missing the information over which stack of the series I'm currently hovering.

Solution:  Get the name of the element accessing the parent's data, and using that name to get the value
https://github.com/AfricasVoices/OperationsDashboard/pull/182/commits/ba5229a175ffe228f529157a5b790d4e2ed98a13

File: public/graph.js [hunk 8]
```diff
@@ -501,7 +502,38 @@ class GraphController {
...
+                        operatorName = operatorReceived.replace('_received',''),
+                        // Get color of hovered rect
+                        operatorColor = d3.select(n[i]).style("fill");
+                    tip = d3.tip()
+                        .attr("class", "tooltip")
+                        .attr("id", "tooltip")
+                        .html(d => { 
+                            let receivedMessages = d.data[operatorReceived],
+                                totalReceivedMessages = d.data.total_received,
+                                receivedDay = d.data.day,
```
Comment (IsaackMwenda) [inline public/graph.js]: no need for the date info because it is already rep in the axes.

**Comment by DanielMwendwa** [comment]
Updated tooltip content by removing date info

File: public/graph.js [hunk 9]
```diff
@@ -450,7 +450,37 @@ class GraphController {
...
+                    let operatorReceived = d3.select(n[i].parentNode).datum().key,
+                        // Get operator name from the key
+                        operatorName = operatorReceived.replace('_received',''),
+                        // Get color of hovered rect
+                        operatorColor = d3.select(n[i]).style("fill");
+                    tip = d3.tip()
+                        .attr("class", "tooltip")
+                        .attr("id", "tooltip")
+                        .html(d => { 
+                            let receivedMessages = d.data[operatorReceived],
```
Comment (IsaackMwenda) [inline public/graph.js]: operatorReceived?

File: public/graph.js [hunk 10]
```diff
@@ -450,7 +450,37 @@ class GraphController {
...
+                                totalReceivedMessages = d.data.total_received,
+                                receivedDay = d.data.day,
```
Comment (IsaackMwenda) [inline public/graph.js]: we no longer need this variable

File: public/graph.js [hunk 11]
```diff
@@ -450,7 +450,37 @@ class GraphController {
...
+                                // Tooltip with operator name, date, no. of msg(s) & msg percentage in that day.
+                                tooltipContent = `<div>${operatorName.charAt(0).toUpperCase() + operatorName.slice(1)}</div>`;
```
Comment (IsaackMwenda) [inline public/graph.js]: 28 (70%) Hormuud Messages?

**Review by IsaackMwenda — APPROVED** [review]
Great this LGTM! @DanielMwendwa. Its also going to be really useful. My only concern now is the duplication of the code once we apply this to outgoing and failed graphs. We might need to apply OOP for graph.js soon. An issue, for now, is okay. Thanks.

File: public/graph.js (hunk 9 above)
Comment (DanielMwendwa) [reply public/graph.js]: I'll update to `operatorNameWithMessageDirection` because it's the operator name concatenated with the msg direction

**Comment by DanielMwendwa** [comment]
Thanks @IsaackMwenda

**Pull request merged**